tt_metal/hw/inc/dataflow_api.h

```diff
@@ -1525,11 +1528,50 @@ void cq_pop_front(uint32_t cmd_size_B) {
     // host and device are consistent in updating their pointers in this way, so they won't get out of sync. The
     // alignment is necessary because we can only read/write from/to 32B aligned addrs in host<->dev communication.
     uint32_t cmd_size_16B = align(cmd_size_B, 32) >> 4;
-    cq_read_interface.fifo_rd_ptr += cmd_size_16B;
+    cq_read_interface.issue_fifo_rd_ptr += cmd_size_16B;
 
     notify_host_of_cq_read_pointer();
 }
 
+FORCE_INLINE
+void cq_reserve_back(uint32_t data_size_B) {
+    DEBUG_STATUS('N', 'Q', 'R', 'B', 'W');
+    uint32_t data_size_16B = align(data_size_B, 32) >> 4;
+    uint32_t rd_ptr_and_toggle;
+    uint32_t rd_ptr;
+    uint32_t rd_toggle;
+    do {
+        rd_ptr_and_toggle = *get_cq_completion_read_ptr();
+        rd_ptr = rd_ptr_and_toggle & 0x7fffffff;
+        rd_toggle = rd_ptr_and_toggle >> 31;
+    } while (
+        cq_write_interface.completion_fifo_wr_ptr < rd_ptr and cq_write_interface.completion_fifo_wr_ptr + data_size_16B > rd_ptr or
+        (rd_toggle != cq_write_interface.completion_fifo_wr_toggle and cq_write_interface.completion_fifo_wr_ptr == rd_ptr)
+    );
+
+    DEBUG_STATUS('N', 'Q', 'R', 'B', 'D');
+}
+
+FORCE_INLINE
+void cq_push_back(uint32_t push_size_B) {
+    uint32_t push_size_16B = align(push_size_B, 32) >> 4;
+    cq_write_interface.completion_fifo_wr_ptr += push_size_16B;
+    if (cq_write_interface.completion_fifo_wr_ptr >= cq_write_interface.completion_fifo_limit) {
+        cq_write_interface.completion_fifo_wr_ptr = DeviceCommand::COMMAND_ISSUE_REGION_SIZE >> 4;
+
+        // Flip the toggle
+        cq_write_interface.completion_fifo_wr_toggle = not cq_write_interface.completion_fifo_wr_toggle;
+    }
+
+    // Notify host of updated completion wr ptr
+    constexpr static uint64_t pcie_address = (uint64_t(NOC_XY_ENCODING(PCIE_NOC_X, PCIE_NOC_Y)) << 32) | HOST_CQ_COMPLETION_WRITE_PTR;  // For now, we are writing to host hugepages at offset
+    uint32_t wr_ptr_and_toggle = cq_write_interface.completion_fifo_wr_ptr | (cq_write_interface.completion_fifo_wr_toggle << 31);
+    volatile tt_l1_ptr uint32_t* wr_ptr_addr = get_cq_completion_write_ptr();
+    wr_ptr_addr[0] = wr_ptr_and_toggle;
+    noc_async_write(CQ_COMPLETION_WRITE_PTR, pcie_address, 4);
+    noc_async_write_barrier();
```

Comment:
I added a write flush barrier that can be used here.

Reply:
do you mean to use `noc_async_writes_flushed()` instead of `noc_async_write_barrier()`?

Reply:
Yes, the noc_async_writes_flushed API just ensures that the data being written has left source, but no guarantees that it has arrived to the destination.

Reply:
don't we want to ensure that it arrived at the dst?

Reply:
We can keep the barrier for now and revisit.